### DISCUSSION TITLE
Message received handler called without actually receiving the SMS

### DISCUSSION THREAD

**Discussion by micartey** [python-smpplib/python-smpplib#244] [Q&A] [answered]
The handler gets called while the SMS is not received.
I tested it by turning flight mode on and thus disconnecting from cell services.

```python
client.set_message_received_handler(lambda pdu: handle_deliver_sm(pdu))
```

`handle_deliver_sm` gets called after some time which appears to be random.

I printed some fields of the `DeliverSM` object in hopes to see a flag or sth similar...
However, I could not find anything that seemed to be different from actually receiving the SMS

```python
client.listen(auto_send_enquire_link=False)
...
client.send_message(
            source_addr_ton=smpplib.consts.SMPP_TON_ALNUM,
            source_addr_npi=smpplib.consts.SMPP_NPI_UNK,
            source_addr=config.SOURCE_NAME,
            dest_addr_ton=smpplib.consts.SMPP_TON_INTL,
            dest_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
            destination_addr=phone_number,
            short_message=part,
            data_coding=encoding_flag,
            esm_class=msg_type_flag,
            registered_delivery=True,
)
```

Am I missing sth?

**Comment by eigenein** [comment]
👋 I'm afraid our example is a little incorrect: `registered_delivery` is an integer not a boolean. In this case `True` is likely sent as `1` which means «MC Delivery Receipt requested where final delivery outcome is delivery **success or failure**». You may consider using `registered_delivery=3` (per SMPP spec 4.7.21)

`DeliverSM` contains the delivery status in the `short_message` and is vendor-specific. So, you may also consider calling `client.query_message` to get the final message state (spec 4.7.15)

**Reply by micartey** [reply]
I changed `registered_delivery` to 3 as you mentioned.
However, I still get the delivered message when I send an SMS to my phone while in airplane mode... 🤔 

The `short_message` states that the message was `delivered` but that can't be true...

**Reply by eigenein** [reply]
This is odd, I'd ask the SMPP service provider about why this happens

**Reply by micartey** [reply, marked as answer]
Turns out we have a SMSC with version 3 and the flag isn't supported until version 5.

Second mistake was that the messages are actually older confirmations that haven't been processed. The SMSC seems to send those confirmations occasionally when not received yet.
